cache/edge_lambdas/src/redirector.ts

```diff
@@ -0,0 +1,89 @@
+import { CloudFrontRequestEvent, CloudFrontResponse } from 'aws-lambda';
+import { URLSearchParams } from 'url';
+import { literalRedirects, queryRedirects } from './redirects';
+
+// This is a more convenient form of the query redirects that looks like:
+// {
+//   [original path to match]: {
+//     params: [URLSearchParams to match]
+//     redirectPath: [path to redirect to]
+//   }
+// }
+const searchParamsRedirects: Map<
+  string,
+  { params: URLSearchParams; redirectPath: string }
+> = new Map();
+for (const fromString in queryRedirects) {
+  const fromUrl = new URL(fromString, 'https://wellcomecollection.org');
+  const params = fromUrl.searchParams;
+  searchParamsRedirects.set(fromUrl.pathname, {
+    params,
+    redirectPath: queryRedirects[fromString],
+  });
+}
+
+const redirect301 = (path: string) => ({
+  status: '301',
+  statusDescription: 'Found',
+  headers: {
+    location: [
+      {
+        key: 'Location',
+        value: `https://wellcomecollection.org${path}`,
+      },
+    ],
+    'x-powered-by': [
+      {
+        key: 'x-powered-by',
+        value: '@weco/redirector',
+      },
+    ],
+  },
+});
+
+// Returns true if and only if all of the params in `subset` are
+// also contained within `superset`
+const paramsAreSubset = (
+  superset: URLSearchParams,
+  subset: URLSearchParams
+): boolean => {
+  for (const pair of subset.entries()) {
+    const [key, value] = pair;
+    if (superset.get(key) !== value) {
+      return false;
+    }
+  }
+  return true;
+};
+
+export const getRedirect = (
```

Comment:
Could you give me an example of this - the comment is a bit abstract for me.

Reply:
So if I make a request for
```
/bananas?foo=bar&right=wrong
```
And in my `queryRedirects` object I have
```
"/bananas?right=wrong": "/plantains"
```
Then a redirect is made to
```
/plantains?foo=bar
```

But, if my redirect was
```
"/bananas?right=wrong&beep=boop": "/plantains"
```
that same request would _not_ redirect because it would not contain the `beep=boop` parameter

Reply:
Is this going to redirect `/works?query=plantains&workType=a,b,c` => `/images?workType=a,b,c`?

Reply:
[P.s.](https://www.youtube.com/watch?v=3SmAlxc7_88)

Reply:
This will redirect `/works?query=plantains&workType=a,b,c&search=images` => `/images?query=plantains&workType=a,b,c`

Reply:
Weirdly the reason I've seen that video before is because @kenoir [shared it](https://wellcome.slack.com/archives/C3TQSF63C/p1574345925013600) in 2019 for reasons now lost to the mists of time

Reply:
Do we want to have it redirect parameters that won't do anything / are invalid?

(the song was one a pal and I wanted to cover in the style of Flamenco)

Reply:
Fair question - I think it's best to leave it for now until/unless we want that feature. The less it knows about the application, the better.

Reply:
I'm with you about it being ignorant, but I think my question is with the abstracted feature. Would you want it to redirect all query params, or ones you have specified. 

I'd argue that given our use case, it's 1 param, that being `query`.

Reply:
Ah, I see what you mean - I had thought it should be all of them in case of "unknown" parameters being useful and because `search=images` did use some filter params that should still be valid for `/images`

Reply:
Which params bar `query` are shared?

Reply:
Params weren't shared between works and images search but `/works?search=images` could also include any of the images params (eg [color](https://github.com/wellcomecollection/wellcomecollection.org/pull/5532/files#diff-800821022eb3268933cdced479d344da7d9c7a178ee9024c7a0629773640628dR129))

Reply:
I think this is partly why we have the current mess with `ImagesRouteProps` and `WorkRouteProps`...

Reply:
Last question on it - but then shouldn't we specify what params we want to keep?

Reply:
Hmmm, yeah I'm kind of on the fence. The intention was that we wouldn't have to know about any of the params except for the one that would 'trigger' the redirect, but I can see the argument for being explicit about them...

Reply:
In my head, if we can avoid creating invalid/wrong URLs we should. We know about the app and the specifics of what we're redirecting. Even if the config is "These are the parameters to redirect" - the redirector code can still be abstract enough.

Reply:
:+1: I'm convinced, will make the change